### PR TITLE
Add the ability to load the KV cache from a file.

### PR DESCRIPTION
**Pull request by angeloskath**
It adds a command called `mlx_lm.cache_prompt` that processes a prompt and saves the KV cache to a file. It also adds the option to load a KV cache from a file in `mlx_lm.generate`.

Combined with long context models it makes for some awesome applications. The following is an example using it to ask question about MLX to Phi-3.5 .

https://github.com/user-attachments/assets/ad386003-0236-4318-b25d-4fa9d7fa8146

### CODE REVIEW

**Comment by fblissjr** [comment]
Is this essentially what Anthropic just released with prompt caching? It looks like it - which is amazing.

**Comment by awni** [comment]
Yea it's the same idea!

**Comment by awni** [comment]
Closes #917

**Comment by angeloskath** [comment]
Added the metadata in the kv cache file. Now we can omit almost all arguments from generate when loading a kv cache.

For instance, below we don't need to pass the model or `max-kv-size` to the generate function, it just grabs them from the kv cache file.
```
cat /tmp/mlx_docs_prompt.txt | mlx_lm.cache_prompt --model mlx-community/Phi-3.5-mini-instruct-8bit --max-kv-size 48000 --prompt "-" --kv-cache-file /tmp/mlx_docs.safetensors
mlx_lm.generate --max-tokens 1024 --kv-cache-file /tmp/mlx_docs.safetensors --prompt "Create a uniform random array with shape 10 by 10 in MLX."
```

**Review by awni — APPROVED** [review]
Awesome! Let's 🚢  it.

**Comment by angeloskath** [comment]
@altaic do you mean at the same time? The way it is set up the `generate` function simply expects a `cache_history` so multiple histories can be loaded in memory and selected on demand (not with the `generate.py` script though).

Loading a new one from a different file is of course trivial.

Also I am not promising anything but @barronalex is looking into quantizing the cache history which will make it a) faster and b) much more sensible to have multiple ones loaded at the same time.

**Comment by ibehnam** [comment]
It looks like this is implemented mainly as a CLI goody, and it's not available in the Python API? 🤔

**Comment by awni** [comment]
Please file an issue if that's something that would be useful for you and we can look into adding it to the Python API.

File: llms/mlx_lm/generate.py [hunk 1]
```diff
@@ -113,6 +118,24 @@ def colorprint_by_t0(s, t0):
...
+    cache_per_layer = {}
+    for k, x in kv_cache.items():
+        layer, kv_type = k.split("_")
+        if layer not in cache_per_layer:
+            cache_per_layer[layer] = {}
+        cache_per_layer[layer][kv_type] = x
+
+    cache_history = [None] * len(cache_per_layer)
+    for layer, c in cache_per_layer.items():
+        cache_history[int(layer)] = (c["keys"], c["values"])
```
Comment (Jonathan-Dobson) [inline llms/mlx_lm/generate.py:135]: Can you fix this type error?

No overloads for "__setitem__" match the provided argumentsPylance[reportCallIssue](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportCallIssue)
builtins.pyi(1028, 9): Overload 1 is the closest match
Argument of type "tuple[Unknown, Unknown]" cannot be assigned to parameter "value" of type "None" in function "__setitem__"
  "tuple[Unknown, Unknown]" is incompatible with "None"Pylance[reportArgumentType](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportArgumentType)


(variable) cache_history: list[None]

**Review by Jonathan-Dobson — COMMENTED** [review]
Can you fix the metadata type error?

Object of type "None" is not subscriptablePylance[reportOptionalSubscript](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportOptionalSubscript)

(variable) metadata: Unknown | None

File: llms/mlx_lm/generate.py [hunk 2]
```diff
@@ -113,6 +118,24 @@ def colorprint_by_t0(s, t0):
     colorprint(color, s)
 
 
+def load_kv_cache_from_file(kv_cache_file):
```
Comment (Jonathan-Dobson) [inline llms/mlx_lm/generate.py:121]: Could you add compatible type definitions for this function.

**Pull request merged**